### PR TITLE
Updates to include references and results by port in report

### PR DESCRIPTION
**Pull request by ididwhat**
As mentioned - update to include references and results by port in word and excel

My first Pyhton code - so hopefully ok.

I haves tested on a range of scans and gives good results.

Hope this helps

iddiwhat

### CODE REVIEW

File: openvasreporting/libs/parsed_data.py [hunk 1]
```diff
@@ -179,7 +205,7 @@ def __init__(self, vuln_id, name, threat, **kwargs):
             raise TypeError("Expected basestring, got '{}' instead".format(type(level)))
         if not isinstance(tags, dict):
             raise TypeError("Expected dict, got '{}' instead".format(type(tags)))
-        if not isinstance(references, list):
+        if not isinstance(references, str):
             raise TypeError("Expected list, got '{}' instead".format(type(references)))
```
Comment (TheGroundZero) [inline openvasreporting/libs/parsed_data.py]: Forgot to update error message after changing variable type:

    raise TypeError("Expected basestring, got '{}' instead".format(type(references)))

File: openvasreporting/libs/parser.py [hunk 2]
```diff
@@ -10,10 +10,10 @@
...
 from .parsed_data import Host, Port, Vulnerability
 
-# logging.basicConfig(stream=sys.stderr, level=logging.DEBUG,
-#                     format="%(asctime)s | %(levelname)s | %(name)s | %(message)s", datefmt="%Y-%m-%d %H:%M:%S")
-logging.basicConfig(stream=sys.stderr, level=logging.ERROR,
-                    format="%(asctime)s | %(levelname)s | %(name)s | %(message)s", datefmt="%Y-%m-%d %H:%M:%S")
+logging.basicConfig(stream=sys.stderr, level=logging.DEBUG,
+                     format="%(asctime)s | %(levelname)s | %(name)s | %(message)s", datefmt="%Y-%m-%d %H:%M:%S")
+#logging.basicConfig(stream=sys.stderr, level=logging.ERROR,
+#                    format="%(asctime)s | %(levelname)s | %(name)s | %(message)s", datefmt="%Y-%m-%d %H:%M:%S")
```
Comment (TheGroundZero) [inline openvasreporting/libs/parser.py]: Don't forget to revert this to prevent unnecessary changes

File: openvasreporting/libs/parser.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def openvas_parser(input_files, min_level=Config.levels()["n"]):
         root = Et.parse(f_file).getroot()
 
         logging.debug("================================================================================")
-        logging.debug("= {}".format(root.find("./task/name").text))  # DEBUG
+#        logging.debug("= {}".format(root.find("./task/name").text))  # DEBUG
```
Comment (TheGroundZero) [inline openvasreporting/libs/parser.py]: Don't forget to revert this to prevent unnecessary changes

File: openvasreporting/libs/parser.py [hunk 4]
```diff
@@ -164,19 +164,35 @@ def openvas_parser(input_files, min_level=Config.levels()["n"]):
...
+
+            logging.debug("* vuln_result:\t{}".format(vuln_result))  # DEBUG
+
             # --------------------
             #
             # STORE VULN_HOSTS PER VULN
             host = Host(vuln_host)
             try:
-                port = Port.string2port(vuln_port)
+	    # added results to port function as will ne unique per port on each host.
```
Comment (TheGroundZero) [inline openvasreporting/libs/parser.py]: ne ~> be ?

File: openvasreporting/libs/parsed_data.py [hunk 5]
```diff
@@ -31,32 +34,45 @@ def __init__(self, number, protocol="tcp"):
...
         self.protocol = protocol
+        self.result = result
 
+    # Modified to include result in structure
     @staticmethod
-    def string2port(info):
+    def string2port(info,result):
         """
         Extract port number, protocol and description from an string.
+	resturn a port class with seperate port, protocol and result
```
Comment (TheGroundZero) [inline openvasreporting/libs/parsed_data.py]:
```
        Extract port number, protocol and description from an string.
	Returns a Port class with separate port, protocol and result.
```

File: openvasreporting/libs/parsed_data.py [hunk 6]
```diff
@@ -31,32 +34,45 @@ def __init__(self, number, protocol="tcp"):
...
             Raises value error if information can't be processed.
 
-        # >>> p=Port.string2port("2000/tcp")
+        # >>> p=Port.string2port("2000/tcp","result string")
         # >>> print p.number
           2000
         # >>> print p.proto
           "tcp"
+	# >>> print p.result
+	  "string"
```
Comment (TheGroundZero) [inline openvasreporting/libs/parsed_data.py]: Looks like this output doesn't match your example input

Shouldn't this be `result string`

File: openvasreporting/libs/parsed_data.py [hunk 7]
```diff
@@ -31,32 +34,45 @@ def __init__(self, number, protocol="tcp"):
...
 
-        # >>> p=Port.string2port("general/icmp")
+        # >>> p=Port.string2port("general/icmp", "string test")
         # >>> print p.number
           0
         # >>> print p.proto
           "icmp"
+	# >>> print p.result
+	  "stringntest"
```
Comment (TheGroundZero) [inline openvasreporting/libs/parsed_data.py]: Looks like this output doesn't match your example input

Shouldn't this be `string test`

File: openvasreporting/libs/export.py [hunk 8]
```diff
@@ -45,12 +51,25 @@ def _get_collections(vuln_info):
...
 
     for i, vuln in enumerate(vuln_info, 1):
         vuln_levels[vuln.level.lower()] += 1
-        vuln_host_by_level[vuln.level.lower()] += len(vuln.hosts)
+        # add host names to list so we count unquie hosts per level
+        level_index = level_choices.get(vuln.level.lower())
+
+        for i, (host, port) in enumerate(vuln.hosts, 1):    
+            if host.ip not in vuln_hostcount_by_level[level_index]:
+                vuln_hostcount_by_level[level_index].append(host.ip)
```
Comment (TheGroundZero) [inline openvasreporting/libs/export.py]: Couldn't you use 

    vuln_hostcount_by_level[vuln.level.lower()].append(host.ip)

Comment (ididwhat) [reply openvasreporting/libs/export.py]: No  vuln_hostcount_by_level is an array of lists so we need the index to be an int and vuln.level is a string, hence the use of the choice clause above, to get an appropriate int.

File: openvasreporting/libs/parser.py (hunk 3 above)
Comment (ididwhat) [reply openvasreporting/libs/parser.py]: agreed - can go back in but this syntax errors when running in Kali.. Hence the reason it was commented out

File: openvasreporting/libs/parser.py (hunk 2 above)
Comment (ididwhat) [reply openvasreporting/libs/parser.py]: agred

File: openvasreporting/libs/parsed_data.py (hunk 1 above)
Comment (ididwhat) [reply openvasreporting/libs/parsed_data.py]: agreed

File: openvasreporting/libs/parser.py (hunk 4 above)
Comment (ididwhat) [reply openvasreporting/libs/parser.py]: agreed

File: openvasreporting/libs/parsed_data.py (hunk 5 above)
Comment (ididwhat) [reply openvasreporting/libs/parsed_data.py]: agreed

File: openvasreporting/libs/parsed_data.py (hunk 6 above)
Comment (ididwhat) [reply openvasreporting/libs/parsed_data.py]: corrected

**Comment by ididwhat** [comment]
Comments addressed in 2 new commits

kr

**Comment by ididwhat** [comment]
updated parser.py to include code to handle missing description tag, this can occur if anon xml is used. so check prevents error in python and code complete ok
